File: typescript/src/_merge.ts
/**
 * Internal deep-merge. Not part of the public surface.
 *
 * - Plain objects recurse key-wise.
 * - Arrays and scalars replace.
 * - Neither input is mutated.
 */

export function deepMerge(
  base: Record<string, unknown>,
  overlay: Record<string, unknown>,
): Record<string, unknown> {
  const result: Record<string, unknown> = { ...base };
  for (const [key, overlayVal] of Object.entries(overlay)) {
    const baseVal = result[key];
    if (isPlainObject(baseVal) && isPlainObject(overlayVal)) {
      result[key] = deepMerge(baseVal, overlayVal);
    } else {
      result[key] = overlayVal;
    }
  }
  return result;
}

export function isPlainObject(v: unknown): v is Record<string, unknown> {
  if (v === null || typeof v !== "object") {
    return false;
  }
  if (Array.isArray(v)) {
    return false;
  }
  const proto: unknown = Object.getPrototypeOf(v);
  return proto === null || proto === Object.prototype;
}

File: python/src/contriwork_config_core/_merge.py
"""Internal deep-merge. Not part of the public surface."""

from __future__ import annotations

from typing import Any


def deep_merge(base: dict[str, Any], overlay: dict[str, Any]) -> dict[str, Any]:
    """Deep-merge two dicts; overlay wins.

    - Dicts recurse key-wise.
    - Lists, scalars, and type mismatches replace.
    - Neither argument is mutated.
    """
    result: dict[str, Any] = dict(base)
    for key, overlay_val in overlay.items():
        base_val = result.get(key)
        if isinstance(base_val, dict) and isinstance(overlay_val, dict):
            result[key] = deep_merge(base_val, overlay_val)
        else:
            result[key] = overlay_val
    return result
